buildtools/npm/manifest.go

```diff
@@ -35,3 +71,96 @@ type Lockfile struct {
 func FromLockfile(filename string) (Lockfile, error) {
 	return Lockfile{}, errors.ErrNotImplemented
 }
+
+// TODO: add support for NODE_PATH and GLOBAL_FOLDERS.
+func modulePath(startingDir string, moduleName string) (string, error) {
+	filePath, err := files.WalkUp(startingDir, func(currentDir string) (err error) {
+		if filepath.Base(currentDir) == moduleName {
+			return files.ErrStopWalk
+		}
+
+		exists, err := files.ExistsFolder(filepath.Join(currentDir, moduleName))
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			return files.ErrStopWalk
+		}
+
+		return nil
+	})
+
+	return filepath.Join(filePath, moduleName, "package.json"), err
+}
+
+func convertManifestToPkg(manifest manifest) pkg.Package {
+	id := pkg.ID{
+		Type:     pkg.NodeJS,
+		Name:     manifest.Name,
+		Revision: manifest.Version,
+	}
+
+	var imports pkg.Imports
+	for depName, version := range manifest.Dependencies {
+		id := pkg.ID{
+			Type:     pkg.NodeJS,
+			Name:     depName,
+			Revision: version,
+		}
+
+		depImport := pkg.Import{
+			Target:   depName,
+			Resolved: id,
+		}
+		imports = append(imports, depImport)
+	}
+
+	return pkg.Package{
+		ID:      id,
+		Imports: imports,
+	}
+}
+
+func fromModulesHelper(pathToModule string) (map[pkg.ID]pkg.Package, error) {
+	moduleProjects := make(map[pkg.ID]pkg.Package)
+	currentDir := filepath.Dir(pathToModule)
+
+	currentModule, err := PackageFromManifest(pathToModule)
+	if err != nil {
+		println(pathToModule)
+		return nil, err
+	}
+
+	for i, imported := range currentModule.Imports {
+		currentDirWithNodeModules := filepath.Join(currentDir, "node_modules")
+		pathToDepModule, err := modulePath(currentDirWithNodeModules, imported.Target)
+		if err != nil {
+			return nil, err
+		}
+
+		modulePackage, err := PackageFromManifest(pathToDepModule) //, "package.json")
+		if err != nil {
+			return nil, err
+		}
+
+		moduleProjects[modulePackage.ID] = modulePackage
+
+		// update previous project's revision resolved reference to stamp out non-deterministic behavior (e.g. semver defined versions in package.json)
+		currentModule.Imports[i].Resolved = modulePackage.ID
+		println("UPDATING " + currentModule.Imports[i].Target + "@" + currentModule.Imports[i].Resolved.Revision + " to " + modulePackage.ID.Revision)
+
+		moduleProjects[currentModule.ID] = currentModule
+
+		nextLevelSubModules, err := fromModulesHelper(pathToDepModule)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, nextLevelSubModule := range nextLevelSubModules {
```

Comment:
[nit] This copying will get `O(scary)`. To avoid it, you can pass a pointer to a map used at all levels of recursion.

Unless you're seeing performance issues right now, I wouldn't worry about it.